mesh/ballotwriter/ballotwriter.go

```diff
@@ -0,0 +1,206 @@
+			w.atxMu.Lock()
+			startTime := time.Now()
+			w.running = false
+			if len(w.ballotBatch) == 0 {
+				w.atxMu.Unlock()
+				continue
+			}
+			batch := w.ballotBatch
+			res := w.ballotBatchResult // copy the result type
+			BatchWriteCount.Inc()
+			FlushBatchSize.Add(float64(len(batch)))
+
+			var ballotAddDur, layerBallotDur time.Duration
+			// we use a context.Background() because: on shutdown the canceling of the
+			// context may exit the transaction halfway and leave the db in some state where it
+			// causes crawshaw to panic on a "not all connections returned to pool".
+			if err := w.db.WithTx(context.Background(), func(tx sql.Transaction) error {
+				for _, ballot := range batch {
+					if !ballot.IsMalicious() {
+						layerBallotStart := time.Now()
+						prev, err := ballots.LayerBallotByNodeID(tx, ballot.Layer, ballot.SmesherID)
+						if err != nil && !errors.Is(err, sql.ErrNotFound) {
+							return err
+						}
+						layerBallotDur += time.Since(layerBallotStart)
+
+						if prev != nil && prev.ID() != ballot.ID() {
+							var ballotProof wire.BallotProof
+							for i, b := range []*types.Ballot{prev, ballot} {
+								ballotProof.Messages[i] = wire.BallotProofMsg{
+									InnerMsg: types.BallotMetadata{
+										Layer:   b.Layer,
+										MsgHash: types.BytesToHash(b.HashInnerBytes()),
+									},
+									Signature: b.Signature,
+									SmesherID: b.SmesherID,
+								}
+							}
+							proof := &wire.MalfeasanceProof{
+								Layer: ballot.Layer,
+								Proof: wire.Proof{
+									Type: wire.MultipleBallots,
+									Data: &ballotProof,
+								},
+							}
+							encoded := codec.MustEncode(proof)
+							if err := identities.SetMalicious(tx, ballot.SmesherID, encoded, time.Now()); err != nil {
+								return fmt.Errorf("add malfeasance proof: %w", err)
+							}
+							ballot.SetMalicious()
+							w.logger.Warn("smesher produced more than one ballot in the same layer",
+								zap.Stringer("smesher", ballot.SmesherID),
+								zap.Object("prev", prev),
+								zap.Object("curr", ballot),
+							)
+						}
+					}
+					ballotAddStart := time.Now()
+					if err := ballots.Add(tx, ballot); err != nil && !errors.Is(err, sql.ErrObjectExists) {
+						return err
+					}
+					ballotAddDur += time.Since(ballotAddStart)
+
+				}
+				return nil
+			}); err != nil {
+				res.err = err
+				ErroredBatchCount.Inc()
+				w.logger.Error("flush ballots to db", zap.Error(err))
+			}
+			cleanupStart := time.Now()
+			w.ballotBatchResult = &batchResult{doneC: make(chan struct{})}
+			for k := range w.ballotBatchIdentities {
+				w.rmNodeSlice = append(w.rmNodeSlice, k)
+			}
+			for _, k := range w.rmNodeSlice {
+				delete(w.ballotBatchIdentities, k)
+			}
+			w.rmNodeSlice = w.rmNodeSlice[:0]
+
+			for k := range batch {
+				w.rmSlice = append(w.rmSlice, k)
+			}
+			for _, k := range w.rmSlice {
+				delete(batch, k)
+			}
+			w.rmSlice = w.rmSlice[:0]
```

Comment:
If I understand correctly, you want to clear `w.ballotBatchIdentities` and `w.ballotBatch`. How about:
```suggestion
			clear(w.ballotBatchIdentities)
			clear(w.ballotBatch)
```

and remove the `rmSlice` and `rmNodeSlice` fields from the `BallotWriter` struct

Reply:
nice. I actually didn't know this builtin